### PR TITLE
Add support for clipping only specific corners, add unit tests

### PR DESCRIPTION
**Pull request by Adlai-Holler**
This does what it says. We add the `maskedCorners` property from iOS 11 and we support it in all of our rounding modes. I added snapshot tests for each case.

### CODE REVIEW

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/6fb485ca806917fcfe3290120e3a2086/raw/bb66e3e9bc2bb02c379063a3380e4bade216132e/log.txt)

**Comment by Adlai-Holler** [comment]
The CI failure indicates that `testEnqueueSameObjectTwiceToNonExclusiveMembershipQueue` is flaky. [log](https://gist.githubusercontent.com/pinterest-ios/6fb485ca806917fcfe3290120e3a2086/raw/bb66e3e9bc2bb02c379063a3380e4bade216132e/log.txt)

**Comment by pinterest-ios** [comment]
🚫 CI failed with [log](https://gist.githubusercontent.com/pinterest-ios/c61caa4729b0067816071ac18552a461/raw/e98f460a740352ed166491753e14499932c33f92/log.txt)

File: Source/ASDisplayNode.mm [hunk 1]
```diff
@@ -1546,78 +1547,87 @@ - (void)_updateClipCornerLayerContentsWithRadius:(CGFloat)radius backgroundColor
...
+      if (_clipCornerLayers[idx] == nil) {
+        continue;
+      }
+
+      // Layers are, in order: Top Left, Top Right, Bottom Left, Bottom Right, which mirrors CACornerMask.
       // anchorPoint is Bottom Left at 0,0 and Top Right at 1,1.
       BOOL isTop   = (idx == 0 || idx == 1);
-      BOOL isRight = (idx == 1 || idx == 2);
-      
+      BOOL isRight = (idx == 1 || idx == 3);
```
Comment (Adlai-Holler) [inline Source/ASDisplayNode.mm]: The reason some of these bits flipped is (1) I changed our NUM_CLIP_CORNERS array order to make the CACornerMask order and (2) the original comments about CA being inverted in this context are incorrect.

File: Source/ASDisplayNode.mm [hunk 2]
```diff
@@ -1546,78 +1547,87 @@ - (void)_updateClipCornerLayerContentsWithRadius:(CGFloat)radius backgroundColor
...
-      }
-      [self _updateClipCornerLayerContentsWithRadius:_cornerRadius backgroundColor:self.backgroundColor];
-    } else {
-      for (int idx = 0; idx < NUM_CLIP_CORNER_LAYERS; idx++) {
+    for (int idx = 0; idx < NUM_CLIP_CORNER_LAYERS; idx++) {
+      BOOL visible = (0 != (visibleCornerLayers & (1 << idx)));
+      if (visible == (_clipCornerLayers[idx] != nil)) {
+        continue;
+      } else if (visible) {
+        static ASDisplayNodeCornerLayerDelegate *clipCornerLayers;
```
Comment (maicki) [inline Source/ASDisplayNode.mm]: We should rename that variable at some point: `clipCornerLayersDelegate`

**Comment by foxware00** [comment]
@Adlai-Holler I would love to use this where i've had to do some messy work arounds to achieve this. Would it be possible to cut a release or is there an easy way for me to use this?

**Pull request merged**